programmer/tinyprog/__init__.py

```diff
@@ -386,6 +467,60 @@ def boot(self):
             # for the boot command data packet.
             pass
 
+    # similar to program_sectors
+    def verify_sectors(self, addr, data, retry=10):
+        possible_lengths = (1, 4 * 1024, 32 * 1024, 64 * 1024)
+        retries_remaining = retry
+        length = len(data)
+        offset = 0
+        write_addr = addr + offset
+        with tqdm(desc="  Verifying", unit="B", unit_scale=True, total=length, disable=False) as pbar:
+            while length > 0 and retries_remaining > 0:
```

Comment:
Verify whether this entire while loop is necessary here.  It appears that this particular loop does nothing.

Reply:
I wanted to have double verify, first verify is sector-by-sector based but
final verify is against complete data. If e.g. sector address was wrong
or overlap for whatever reason, then each sector will verify OK but final
result won't be OK.

Reply:
Take a closer look at the while loop line by line...it's not actually doing anything.  It just updates local variables and updates the progress bar.
```python
            while length > 0 and retries_remaining > 0:
                erase_length = max(p for p in possible_lengths
                                   if p <= length and write_addr % p == 0)
                if erase_length == 1:
                    # there are no opcode to erase that much
                    # either we want to erase up to multiple of 0x1000
                    # or we want to erase up to length

                    # start_addr                            end_addr
                    # v                                     v
                    # +------------------+------------------+----------------+
                    # |       keep       |      erase       |      keep      |
                    # +------------------+------------------+----------------+
                    #  <- start_length -> <- erase_length -> <- end_length ->

                    start_addr = write_addr & 0xfff000
                    start_length = write_addr & 0xfff
                    erase_length = min(0x1000 - start_length, length)
                    end_addr = start_addr + start_length + erase_length
                    end_length = start_addr + 0x1000 - end_addr

                self.progress(erase_length)

                write_data = data[offset : offset + erase_length]
                # read_back = self.read(write_addr, erase_length)
                read_back = write_data

                # update
                if read_back == write_data:
                  length -= erase_length
                  write_addr += erase_length
                  offset += erase_length
                  retries_remaining = retry
                  pbar.update(erase_length)
                else:
                  retries_remaining -= 1
```

Reply:
Maybe you are right :) I had to uncomment
# read_back = self.read.bla
Let's say a verify-only commandline option would be good to have